File: aoc2021/aoc4.hs
import System.Environment

every :: Int -> [a] -> [a]
every _ [] = []
every n lst = head lst : every n (drop n lst)

won :: [Int] -> [Int] -> Bool
won board called =
    let boardfound = [if x `elem` called then 1 else 0 | x <- board] :: [Int]
        horizwon = any (\idx -> 5 == sum (take 5 $ drop (5*idx) boardfound)) [0..4]
        vertwon = any (\idx -> 5 == sum (every 5 $ drop idx boardfound)) [0..4]
    in horizwon || vertwon

score :: (Num a, Eq a) => [a] -> [a] -> a
score board called =
    let lastcalled = last called
        unmarked = [x | x <- board, x `notElem` called]
    in lastcalled * sum unmarked

makeboards :: [String] -> [[Int]]
makeboards [] = []
makeboards s = map read (take 25 s) : makeboards (drop 25 s)

p1 :: [[Int]] -> [Int] -> Int
p1 boards called =
    head [score b (take idx called) | idx <- [4..length called], b <- boards, won b (take idx called)]

p2 :: [[Int]] -> [Int] -> Int
p2 boards' called = uncurry score $ go 4 boards'
    where
        go :: Int -> [[Int]] -> ([Int], [Int])
        go _ [] = error "empty boards"
        go idx [a] = if won a (take idx called) then (a, take idx called)
                     else go (idx+1) [a]
        go idx boards = let sofar = take idx called
                            remaining = [b | b <- boards, not (won b sofar)]
                        in go (idx+1) remaining

main :: IO ()
main = do
  args <- getArgs
  let filename = if null args then "aoc4.in" else head args
  strs <- words <$> readFile filename
  let called = read ("[" ++ head strs ++ "]") :: [Int]
  let boards = makeboards (tail strs)
  print $ p1 boards called
  print $ p2 boards called
  

File: aoc2021/aoc4.py
import sys
import re
import itertools

num_called = []
boards = []
with open('aoc4.in' if len(sys.argv) < 2 else sys.argv[1]) as f:
    data = re.findall(r'\S+', f.read())
    num_called = [int(x) for x in data[0].split(",")]
    data = data[1:]
    for start_idx in range(0, len(data), 25):
        boards.append([int(x) for x in data[start_idx:start_idx+25]])

def won(board, called):
    cset = set(called)
    bc = [int(val in cset) for val in board]
    for horiz in range(5):
        if sum(bc[horiz*5:horiz*5+5]) == 5:
            return True
    for vert in range(5):
        if bc[vert] + bc[vert + 5] + bc[vert + 10] + bc[vert + 15] + bc[vert + 20] == 5:
            return True
    return False

def score(board, called):
    cset = set(called)
    uncalled = sum([val for val in board if val not in cset])
    last_called = called[-1]
    return last_called * uncalled


done = False
for game_idx in range(4, len(num_called)):
    sofar = num_called[0:game_idx]
    for board in boards:
        if won(board, sofar):
            print(score(board, sofar))
            done = True
            break
    if done:
        break


done = False
workboards = boards
for game_idx in range(4, len(num_called)):
    sofar = num_called[0:game_idx]
    for (idx, board) in list(reversed(list(enumerate(workboards)))):
        if won(board, sofar):
            if len(workboards) == 1:
                print(score(board, sofar))
                done = True
                break
            del workboards[idx]
    if done:
        break
